File: src/fable-library/Option.ts
import { structuralHash, equals, compare } from "./Util.js";

// Options are erased in runtime by Fable, but we have
// the `Some` type below to wrap values that would evaluate
// to `undefined` in runtime. These two rules must be followed:

// 1- `None` is always `undefined` in runtime, a non-strict null check
//    (`x == null`) is enough to check the case of an option.
// 2- To get the value of an option the `value` helper
//    below must **always** be used.

// Note: We use non-strict null check for backwards compatibility with
// code that use F# options to represent values that could be null in JS

export type Nullable<T> = T | null | undefined;

export type Option<T> = T | Some<T> | undefined;

// Using a class here for better compatibility with TS files importing Some
export class Some<T> {
  public value: T;

  constructor(value: T) {
    this.value = value;
  }

  public toJSON() {
    return this.value;
  }

  // Don't add "Some" for consistency with erased options
  public toString() {
    return String(this.value);
  }

  public GetHashCode() {
    return structuralHash(this.value);
  }

  public Equals(other: Option<T>): boolean {
    if (other == null) {
      return false;
    } else {
      return equals(this.value, other instanceof Some ? other.value : other);
    }
  }

  public CompareTo(other: Option<T>) {
    if (other == null) {
      return 1;
    } else {
      return compare(this.value, other instanceof Some ? other.value : other);
    }
  }
}

export function value<T>(x: Option<T>) {
  if (x == null) {
    throw new Error("Option has no value");
  } else {
    return x instanceof Some ? x.value : x;
  }
}

export function unwrap<T>(opt: Option<T>): T | undefined {
  return opt instanceof Some ? opt.value : opt;
}

export function some<T>(x: T): Option<T> {
  return x == null || x instanceof Some ? new Some(x) : x;
}

export function ofNullable<T>(x: Nullable<T>): Option<T> {
  // This will fail with unit probably, an alternative would be:
  // return x === null ? undefined : (x === undefined ? new Some(x) : x);
  return x == null ? undefined : x;
}

export function toNullable<T>(x: Option<T>): Nullable<T> {
  return x == null ? null : value(x);
}

export function flatten<T>(x: Option<Option<T>>) {
  return x == null ? undefined : value(x);
}

export function toArray<T>(opt: Option<T>): T[] {
  return (opt == null) ? [] : [value(opt)];
}

export function defaultArg<T>(opt: Option<T>, defaultValue: T): T {
  return (opt != null) ? value(opt) : defaultValue;
}

export function defaultArgWith<T>(opt: Option<T>, defThunk: () => T): T {
  return (opt != null) ? value(opt) : defThunk();
}

export function orElse<T>(opt: Option<T>, ifNone: Option<T>): Option<T> {
  return opt == null ? ifNone : opt;
}

export function orElseWith<T>(opt: Option<T>, ifNoneThunk: () => Option<T>): Option<T> {
  return opt == null ? ifNoneThunk() : opt;
}

export function filter<T>(predicate: (arg: T) => boolean, opt: Option<T>): Option<T> {
  return (opt != null) ? (predicate(value(opt)) ? opt : undefined) : opt;
}

export function map<T, U>(mapping: (arg: T) => U, opt: Option<T>): Option<U> {
  return (opt != null) ? some(mapping(value(opt))) : undefined;
}

export function map2<T1, T2, U>(
  mapping: (arg1: T1, arg2: T2) => Option<U>,
  opt1: Option<T1>, opt2: Option<T2>): Option<U> {
  return (opt1 != null && opt2 != null) ? mapping(value(opt1), value(opt2)) : undefined;
}

export function map3<T1, T2, T3, U>(
  mapping: (arg1: T1, arg2: T2, arg3: T3) => Option<U>,
  opt1: Option<T1>, opt2: Option<T2>, opt3: Option<T3>): Option<U> {
  return (opt1 != null && opt2 != null && opt3 != null) ? mapping(value(opt1), value(opt2), value(opt3)) : undefined;
}

export function bind<T, U>(binder: (arg: T) => Option<U>, opt: Option<T>): Option<U> {
  return opt != null ? binder(value(opt)) : undefined;
}

export function tryOp<T, U>(op: (x: T) => U, arg: T): Option<U> {
  try {
    return some(op(arg));
  } catch {
    return undefined;
  }
}

File: src/fable-library-dart/Option.fs
module OptionModule

open System

let defaultValue (def: 'T) (opt: 'T option) : 'T =
    match opt with
    | None -> def
    | Some opt -> opt

let defaultWith (fn: unit -> 'T) (opt: 'T option) : 'T =
    match opt with
    | None -> fn ()
    | Some opt -> opt

let orElse (def: 'T option) (opt: 'T option) : 'T option =
    match opt with
    | None -> def
    | Some opt -> Some opt

let orElseWith (fn: unit -> 'T option) (opt: 'T option) : 'T option =
    match opt with
    | None -> fn ()
    | Some opt -> Some opt

let toArray (opt: 'T option) : 'T[] =
    match opt with
    | None -> [||]
    | Some opt -> [| opt |]

let toList (opt: 'T option) : 'T list =
    match opt with
    | None -> []
    | Some opt -> [ opt ]

let count (opt: 'T option) : int =
    match opt with
    | Some _ -> 1
    | None -> 0

let contains (value: 'T) (opt: 'T option) : bool =
    match opt with
    | Some value2 -> value = value2
    | None -> false

let iterate (fn: 'T -> unit) (opt: 'T option) : unit =
    match opt with
    | Some opt -> fn opt
    | None -> ()

let forAll (fn: 'T -> bool) (opt: 'T option) : bool =
    match opt with
    | Some opt -> fn opt
    | None -> true

let exists (fn: 'T -> bool) (opt: 'T option) : bool =
    match opt with
    | Some opt -> fn opt
    | None -> false

let fold<'T, 'State>
    (folder: 'State -> 'T -> 'State)
    (state: 'State)
    (opt: 'T option)
    : 'State
    =
    match opt with
    | Some opt -> folder state opt
    | None -> state

let foldBack<'T, 'State>
    (folder: 'T -> 'State -> 'State)
    (opt: 'T option)
    (state: 'State)
    : 'State
    =
    match opt with
    | Some opt -> folder opt state
    | None -> state

let filter (fn: 'T -> bool) (opt: 'T option) : 'T option =
    match opt with
    | None -> None
    | Some opt ->
        if fn opt then
            Some opt
        else
            None

let flatten<'T> (opt: 'T option option) : 'T option =
    match opt with
    | Some x -> x
    | None -> None

let map (fn: 'T -> 'U) (opt: 'T option) : 'U option =
    match opt with
    | None -> None
    | Some opt -> fn opt |> Some

let map2
    (fn: 'T1 -> 'T2 -> 'U)
    (opt1: 'T1 option)
    (opt2: 'T2 option)
    : 'U option
    =
    match opt1, opt2 with
    | Some opt1, Some opt2 -> fn opt1 opt2 |> Some
    | _ -> None

let map3
    (fn: 'T1 -> 'T2 -> 'T3 -> 'U)
    (opt1: 'T1 option)
    (opt2: 'T2 option)
    (opt3: 'T3 option)
    : 'U option
    =
    match opt1, opt2, opt3 with
    | Some opt1, Some opt2, Some opt3 -> fn opt1 opt2 opt3 |> Some
    | _ -> None

let bind (fn: 'T -> 'U option) (opt: 'T option) : 'U option =
    match opt with
    | None -> None
    | Some opt -> fn opt

let ofNullable (x: Nullable<'T>) : 'T option =
    if x.HasValue then
        Some(x.Value)
    else
        None

let toNullable (opt: 'T option) : Nullable<'T> =
    match opt with
    | Some x -> Nullable x
    | None -> Nullable()
